### DISCUSSION TITLE
CaffeineProxyManager vs Simple Caffine Based LoadingCache

### DISCUSSION THREAD

**Discussion by supercharger** [vladimir-bukhtoyarov/bucket4j#442] [General]
What's the difference between CaffeineProxyManager vs simple Caffine's LoadingCache equivalent?  Is there any performance difference between the two, because I see [RemoteBucketState/MutableBucketEntry](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-caffeine/src/main/java/io/github/bucket4j/caffeine/CaffeineProxyManager.java#L93) and What do you think about following?

```
public class SessionRateLimiter {
    private final LoadingCache<String, Bucket> ipRateLimits;
    
    private final Duration EXPIRE_AFTER_BUFFER = Duration.ofMinutes(10);

    public SessionRateLimiter(long capacity,  Duration refillDuration) {
    	ipRateLimits = CacheBuilder.newBuilder()
                .expireAfterAccess(EXPIRE_AFTER_BUFFER.plus(refillDuration))
                .build(new CacheLoader<String, Bucket>() {
                    @Override
                    public Bucket load(String key) {
                        return Bucket.builder()
                         .addLimit(limit -> limit.capacity(capacity).refillGreedy(capacity, refillDuration))
                         .build();
                    }
                });
    }

    public void tryAcquire(String ip) {
      Bucket bucket = ipRateLimits.get(ip);
      return bucket.tryConsume(1);
    }
    
    
}
```

**Comment by vladimir-bukhtoyarov** [comment]
Hello @supercharger 

CaffeineProxyManager tries to avoid storing buckets more than it actually needs, as well prevents removing bucket from cache too early when bucket is not refilled up to high. See [this](https://github.com/bucket4j/bucket4j/blob/8.7/bucket4j-caffeine/src/main/java/io/github/bucket4j/caffeine/CaffeineProxyManager.java#L69) and [this](https://github.com/bucket4j/bucket4j/blob/8.7/bucket4j-caffeine/src/main/java/io/github/bucket4j/caffeine/CaffeineProxyManager.java#L70) lines. While your example has two dangerious points:
* It stores bucket for 10 minutes even if actual time to refill bucket up to high is small, for example a few seconds.
* If refill period is huge like 1 hour, it brings the risk of overconsuming, because empty bucket can be forget too early and then ressurected inside `load` as totally refilled up to high.
